### PR TITLE
Change the logic around breaking multiple patterns in match arms

### LINKED ISSUE #386: Option to compress pattern options if there are all simple
```
-        BiEq | BiLt | BiLe | BiNe | BiGt | BiGe =>
-            true,
-        BiAnd | BiOr | BiAdd | BiSub | BiMul | BiDiv | BiRem |
-        BiBitXor | BiBitAnd | BiBitOr | BiShl | BiShr =>
-            false,
+        BiEq | BiLt | BiLe | BiNe | BiGt | BiGe => true,
+        BiAnd |
+        BiOr |
+        BiAdd |
+        BiSub |
+        BiMul |
+        BiDiv |
+        BiRem |
+        BiBitXor |
+        BiBitAnd |
+        BiBitOr |
+        BiShl |
+        BiShr => false,
```

is kind of silly, in cases like this where all the patterns are trivial (that is a single ident with no params), we should use Mixed layout, rather than Vertical layout.

### PR DESCRIPTION
**Pull request by nrc**
Refactor to use the list code, don't preserve original stacking-ness, base vertical vs mixed formatting on complexity of the patterns.

Closes #386 

r? @marcusklaas

### CODE REVIEW

File: src/expr.rs [hunk 1]
```diff
@@ -980,38 +980,34 @@ impl Rewrite for ast::Arm {
...
+            width: pat_budget,
+            ends_with_newline: false,
+            config: context.config,
+        };
+        let pats_str = try_opt!(write_list(items, &fmt));
 
-        let pats_width = if vertical {
-            pat_strs.last().unwrap().len()
+        let budget = if pats_str.contains('\n') {
+            context.config.max_width
```
Comment (marcusklaas) [inline src/expr.rs]: This doesn't look quite correct. Generally speaking, budgets do not include indentation. I think this should be `max_width - offset.width()`.

File: src/config.rs [hunk 2]
```diff
@@ -87,7 +87,8 @@ impl Density {
     pub fn to_list_tactic(self) -> ListTactic {
         match self {
             Density::Compressed => ListTactic::Mixed,
-            Density::Tall | Density::CompressedIfEmpty => ListTactic::HorizontalVertical,
```
Comment (marcusklaas) [inline src/config.rs]: Why isn't this on a single line any more?

**Comment by marcusklaas** [comment]
Looks good to me. I'm a bit hesitant about throwing out the old stacking heuristics, though. Or were those basically "preserve stacking-ness, unless running over column limit"?

File: src/expr.rs [hunk 3]
```diff
@@ -1085,6 +1081,11 @@ impl Rewrite for ast::Arm {
     }
 }
 
+fn pat_is_simple(pat_str: &str) -> bool {
+    pat_str.len() <= 16 ||
+    (pat_str.len() <= 24 && pat_str.chars().all(|c| c.is_alphabetic() || c == ':'))
```
Comment (marcusklaas) [inline src/expr.rs]: What is the alphabet check for? To detect generics?

**Comment by marcusklaas** [comment]
r+ with the width calculation addressed and a second thought on the heuristics.

File: src/expr.rs (hunk 3 above)
Comment (nrc) [inline src/expr.rs]: I'll add a comment, but simple here means a unitary variant (or a variable), e.g., `Foo` or `Foo::Bar`, but excluding tuple patterns, struct patterns, variables with `@`, etc.

File: src/config.rs (hunk 2 above)
Comment (nrc) [inline src/config.rs]: The idea is to put simple patterns on the same line (or mixed), so something `Foo | Bar | Baz` doesn't end up like

```
Foo |
Bar |
Baz
```

**Comment by nrc** [comment]
With the changes here, we're much more likely to have a vertical layout than before, so I don't think we need to keep a heuristic which preserves that layout. The only time we would change vertical to horizontal is for *very* simple patterns, where I think they should never be vertical.

File: src/expr.rs (hunk 1 above)
Comment (nrc) [inline src/expr.rs]: Hmm, I think it is correct here. rewrite_guard is a bit odd in that the indent is accounted for separately in the pat_width arg. But that means that rewrite_arg will go wrong if the guard goes over multiple lines. Let me try and rejig things...

**Pull request merged**